mllib/src/main/scala/org/apache/spark/ml/feature/VectorSizeHint.scala

```diff
@@ -0,0 +1,151 @@
+ */
+@Experimental
+@Since("2.3.0")
+class VectorSizeHint @Since("2.3.0") (@Since("2.3.0") override val uid: String)
+  extends Transformer with HasInputCol with HasHandleInvalid with DefaultParamsWritable {
+
+  @Since("2.3.0")
+  def this() = this(Identifiable.randomUID("vectSizeHint"))
+
+  @Since("2.3.0")
+  val size = new IntParam(this, "size", "Size of vectors in column.", {s: Int => s >= 0})
+
+  @Since("2.3.0")
+  def getSize: Int = getOrDefault(size)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setSize(value: Int): this.type = set(size, value)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  @Since("2.3.0")
+  override val handleInvalid: Param[String] = new Param[String](
+    this,
+    "handleInvalid",
+    "How to handle invalid vectors in inputCol, (invalid vectors include nulls and vectors with " +
+      "the wrong size. The options are `skip` (filter out rows with invalid vectors), `error` " +
+      "(throw an error) and `optimistic` (don't check the vector size).",
+    ParamValidators.inArray(VectorSizeHint.supportedHandleInvalids))
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+  setDefault(handleInvalid, VectorSizeHint.ERROR_INVALID)
+
+  @Since("2.3.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val localInputCol = getInputCol
+    val localSize = getSize
+    val localHandleInvalid = getHandleInvalid
+
+    val group = AttributeGroup.fromStructField(dataset.schema(localInputCol))
+    if (localHandleInvalid == VectorSizeHint.OPTIMISTIC_INVALID && group.size == localSize) {
+      dataset.toDF
+    } else {
+      val newGroup = group.size match {
+        case `localSize` => group
+        case -1 => new AttributeGroup(localInputCol, localSize)
+        case _ =>
+          val msg = s"Trying to set size of vectors in `$localInputCol` to $localSize but size " +
+            s"already set to ${group.size}."
+          throw new SparkException(msg)
+      }
+
+      val newCol: Column = localHandleInvalid match {
+        case VectorSizeHint.OPTIMISTIC_INVALID => col(localInputCol)
+        case VectorSizeHint.ERROR_INVALID =>
+          val checkVectorSizeUDF = udf { vector: Vector =>
+            if (vector == null) {
+              throw new SparkException(s"Got null vector in VectorSizeHint, set `handleInvalid` " +
+                s"to 'skip' to filter invalid rows.")
+            }
+            if (vector.size != localSize) {
+              throw new SparkException(s"VectorSizeHint Expecting a vector of size $localSize but" +
+                s" got ${vector.size}")
+            }
+            vector
+          }.asNondeterministic
+          checkVectorSizeUDF(col(localInputCol))
+        case VectorSizeHint.SKIP_INVALID =>
+          val checkVectorSizeUDF = udf { vector: Vector =>
```

Comment:
Something like:
```
val isSparse = col(localInputCol)(0) === lit(0)
val sparseSize = col(localInputCol)(1)
val denseSize = size(col(localInputCol)(3))
val vecSize = when(isSparse, sparseSize).otherwise(denseSize)
val sizeMatches = vecSize === lit(localSize)
when(col(localInputCol).isNotNull && sizeMatches,
  col(localInputCol),
  lit(null))
```

That should be 90% correct I think  : )

Reply:
Or not?  Maybe the analyzer won't allow you to treat a UDF column as a generic struct.  Scratch this comment.

Reply:
But it doesn't work... UserDefinedType column cannot be used as `StructType`. @cloud-fan Is there any way we can directly extract "Struct" from UDT column ? (in pure sql way)

Reply:
internally UDT column is stored as `UserDefinedType.sqlType`, so if your UDT is mapped to sql struct type, we can use it as struct type column via pure SQL/DataFrame operations.

Reply:
@cloud-fan I tried, but got such error:
`org.apache.spark.sql.AnalysisException: Can't extract value from a#3: need struct type but got vector`, anywhere wrong ? test code:
```
import spark.implicits._
import org.apache.spark.ml.linalg._
val df1 = Seq(Tuple1(Vectors.dense(1.0, 2.0))).toDF("a")
df1.select(col("a")(0)).show
```

Reply:
I feel this is a bug, let me look into it